### PR TITLE
common lib

### PR DESCRIPTION
**Pull request by Gusabary**
Common lib for c++ provider and client, basically
+ it uses proto as submodule and holds logic to generate pb files so provider and client no longer need to add proto as submodule
+ it includes some common components such as `EventLoop`, `NetworkInterface` and so on which could be used by both provider and client
+ it includes some game-specific utils like `MsgBuilder` which could be used by both provider and client of the same game

note that this repo may contain sources of multiple games, before `add_subdirectory(CommonLib)`, choose lib of which game you want to build. e.g. `set(BUILD_UNO_LIB ON)` for uno.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+
+set(protobuf_MODULE_COMPATIBLE TRUE)
+find_package(Protobuf CONFIG REQUIRED)
+message(STATUS "Using protobuf ${Protobuf_VERSION}")
+
+set(_PROTOBUF_LIBPROTOBUF protobuf::libprotobuf)
+set(_REFLECTION gRPC::grpc++_reflection)
+set(_PROTOBUF_PROTOC $<TARGET_FILE:protobuf::protoc>)
+
+find_package(gRPC CONFIG REQUIRED)
```
Comment (Veiasai) [inline CMakeLists.txt]: make this support both install grpc and submodule grpc.
https://github.com/grpc/grpc/blob/master/examples/cpp/helloworld/CMakeLists.txt

File: CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    --cpp_out "${CMAKE_BINARY_DIR}"
+    -I "${proto_dir}"
+    --plugin=protoc-gen-grpc="${_GRPC_CPP_PLUGIN_EXECUTABLE}"
+    # here should be absolute path
+    ${proto_files}
+    DEPENDS "${proto_files}"# "${_PROTOBUF_PROTOC}" "${_GRPC_CPP_PLUGIN_EXECUTABLE}"
+)
+set_source_files_properties(${pb_srcs} PROPERTIES GENERATED TRUE)
+
+add_library(uno_pb_lib ${pb_srcs})
```
Comment (Veiasai) [inline CMakeLists.txt]: should put this in if(BUILD_UNO_LIB)?  or use a variable name?

File: CMakeLists.txt [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+set(pb_dir "${CMAKE_BINARY_DIR}")
+file(GLOB pb_hdrs ${pb_dir}/core/*.pb.h)
+file(GLOB pb_srcs ${pb_dir}/core/*.pb.cc)
+if(BUILD_UNO_LIB)
+    file(GLOB uno_pb_hdrs ${pb_dir}/uno/*.pb.h)
+    file(GLOB uno_pb_srcs ${pb_dir}/uno/*.pb.cc)
+    list(APPEND pb_hdrs ${uno_pb_hdrs})
+    list(APPEND pb_srcs ${uno_pb_srcs})
+endif()
+if(BUILD_UNO_LIB)
```
Comment (Veiasai) [inline CMakeLists.txt]: ???

File: CMakeLists.txt (hunk 1 above)
Comment (Gusabary) [reply CMakeLists.txt]: ok

File: CMakeLists.txt [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+    -I "${proto_dir}"
+    --plugin=protoc-gen-grpc="${_GRPC_CPP_PLUGIN_EXECUTABLE}"
+    # here should be absolute path
+    ${proto_files}
+    DEPENDS "${proto_files}"# "${_PROTOBUF_PROTOC}" "${_GRPC_CPP_PLUGIN_EXECUTABLE}"
+)
+set_source_files_properties(${pb_srcs} PROPERTIES GENERATED TRUE)
+
+add_library(uno_pb_lib ${pb_srcs})
+target_include_directories(uno_pb_lib PUBLIC ${pb_dir})
```
Comment (Veiasai) [inline CMakeLists.txt]: I feel it's better to implement
`#include <sailgame/common/xx>`
`#include <sailgame/common/core/xx`  pb

common is vulnerable...

just add a sailgame dir?

File: CMakeLists.txt (hunk 2 above)
Comment (Gusabary) [reply CMakeLists.txt]: I am gonna use another variable name

File: CMakeLists.txt (hunk 3 above)
Comment (Gusabary) [reply CMakeLists.txt]: typo

File: CMakeLists.txt (hunk 4 above)
Comment (Gusabary) [reply CMakeLists.txt]: yep, that's a good idea

File: include/common/network_interface.h [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+        return GameCore::NewStub(channel);
+    }
+
+    static std::shared_ptr<NetworkInterface> Create(
+        const std::shared_ptr<GameCore::StubInterface> &stub) {
+        return std::make_shared<NetworkInterface>(stub);
+    }
+
+    void Connect() {
+        mStream = mStub->Provider(&mContext);
```
Comment (Veiasai) [inline include/common/network_interface.h]: could you support client workflow?
client ->  regsiter -> listen.

I think 
1. you can make a pure virtual base class, and offer two implementations. one for provider, another for client.
2. Or, just implement two classes, don't use inheritance.

**Review by Veiasai — CHANGES_REQUESTED** [review]

File: include/common/network_interface.h (hunk 5 above)
Comment (Gusabary) [reply include/common/network_interface.h]: sure, that's under design now

**Comment by Gusabary** [comment]
`NetworkInterface` with client side support could be added in next pr.
pls merge this for integration with other providers

**Review by Veiasai — APPROVED** [review]
nice! nice! nice!

**Pull request merged**